### ISSUE TITLE
Map cannot be rotated while zooming

### ISSUE THREAD

**Issue by mcliquid** [streetcomplete/StreetComplete#5844]
Comparison between version 58.2 and 59.0 alpha.
* In version 58, the map can be zoomed and rotated at the same time. No matter which gesture you start with, the map rotates and zooms within the same gesture command.
* In version 59, the map can no longer be rotated while zooming. If you manage to rotate first, then you can zoom. However, if you zoom in first, the map cannot be rotated at the same time.

To make it clearer if needed: Both functions are working fine in v59 but not at the same time in the same gesture.

At least I find it difficult with only one hand to get the map to rotate first and then zoom, my mobile phone reacts to the zoom first in most cases with one hand. It works better with two hands.
(I took screenshots, but this is just a still and not helpful I guess.)

**How to Reproduce**
* Start version 59
* Place two fingers on the display using the common "zoom" gesture
* Zoom in or out and try to rotate the map (in the same gesture/without changing the gesture)
* (Try the same in version 58)

**Expected Behavior**
The map rotates at the same time as the zoom.

**Versions affected**
Version 59.0 alpha (SC and SCEE)

**Comment by westnordost** [comment]
Not sure about v58, but in v59 (MapLibre), this is deliberate. There is a specific option to turn on this behavior, and it has been requested that this should be turned on in the past.

**Comment by westnordost** [comment]
I don't remember. You can try looking through the discussion(s) in the PRs for integrating MapLibre.

**Comment by westnordost** [comment]
#3022

**Comment by westnordost** [comment]
They didn't want removed that you can rotate at all, that's how I understood it.

**Comment by westnordost** [comment]
Anyway, OrganicMaps for example doesn't allow rotating the map while zooming (but it allows zooming while rotating the map). This makes sense. Usually, when zooming, you don't want to rotate the map. While when you rotate, maybe you want to zoom a bit, too.

The behavior in StreetComplete v59.0-alpha1 is exactly the same as in OrganicMaps.

**Comment by westnordost** [comment]
My version is 2024.08.16-5-FDroid but I doubt that they changed it within one day. I can definitely zoom and rotate at the same time. But If I start it as a zoom gesture, I cannot rotate. When I start it as a rotate gesture, I can do both. Same as in StreetComplete v59.0-alpha1

**Comment by westnordost** [comment]
> Well what can I say, you can see what I do in my video, can't you?

Yes, you start with rotating.

**Comment by mnalis** [comment]
- (ref. related https://github.com/streetcomplete/StreetComplete/issues/5836)

**Comment by laundmo** [comment]
I strongly request undoing what to me makes StreetComplete MUCH more annoying to navigate.

Theres also a notable difference to Organic Maps: in StreetComplete, navigating and especially rotating around is much more meccessary than when, for example, following a route.

In any case, if you don't undo this downgrade, i'll jave to request it from SCEE which i'm already using anyways.

You are clearly getting multiple people who want to be able to rotate/zoom freely while only some want the setting to change the navigation to this. At least expose it as a setting so those who are annoyed by this can re-enable it.

**Comment by kmpoppe** [comment]
I don't really see a problem with one gesture only doing zoom and one doing rotation and zoom - BUT 

a) I have a relatively new smartphone with no problems to decide between the two 

b) I don't have any condition that makes it harder for me to precisely move my fingers (i.e. arthritis)

c) I get annoyed by almost any map rotation, I'm used to top-north physical maps.

YET SC has behaved to accept both gestures at the same time forever, changing it now is what my software company would call a breaking change. Introducing a breaking change that is not economically necessary to keep product and customer alive is fine, but always needs to be reversible.

I strongly vouch for an option button ("Allow map rotation and zoom at the same time", don't even phrase it as working in one way but not the other, that's been discussed at length here) and be done. So everyone can have it their way without going for SCEE.

K

**Comment by matkoniecz** [comment]
> Introducing a breaking change that is not economically necessary to keep product and customer alive is fine, but always needs to be reversible.

This general claim is much too strong, often breaking changes are impossible to avoid. Or not reasonable to avoid.

**Comment by laundmo** [comment]
as it very much sounds like it was a concious decision to as this being behavior i don't think it matters much that its impossible to avoid.

>  v59 (MapLibre), this is deliberate. There is a specific option to turn on this behavior, and it has been requested that this should be turned on in the past.

Just, allow users to change this. Don't dictate how users should navigate the map. Let them decide. Choose a default however you want, though i would advocate for the default being the old behaviour to stop this from being a breaking change.

Its such a wierd small thing to be strict about. Different people have different preferences.

If it sounds like i'm too passionate about this: kinda, yeah. I only started using StreetComplete recently, and after using it a LOT (top 50 worldwide last 7 days) i stumble across a crash in the old version, switch to the beta, and am faced with a annoying change only to find out it was on purpose due to 1 person requesting it in the past.

**Comment by laundmo** [comment]
> this is necessary to some degree - alternative is having more settings than Osmand have and maintaining it.

Then leave the behaviour people got used to in place.

Don't go making entirely optional breaking changes if thats the issue.

**Comment by westnordost** [comment]
So, the current (new) behavior is that the map cannot be rotated when the zoom gesture has already been started first (but the other way round). This is new behavior in StreetComplete (since v59) but standard behavior in OrganicMaps.

I too noticed that the thresholds for which gesture is detected are a bit different in MapLibre and OrganicMaps, in the latter it feels more natural while in MapLibre rarely I accidentally zoom even though I wanted to rotate. This is the issue that some users in this thread are mainly complaining about. Formerly, I thought there was nothing that could be done about that other than turning off that behavior completely.

However, recently while investigating another MapLibre bug, I found out that the thresholds for when a two-finger gesture is detected as either shove, rotate or zoom can be adjusted after all. They can be adjusted here:

https://github.com/streetcomplete/StreetComplete/blob/7c04687816c9e455ccedce7858f61cb0d59b6d5d/app/src/main/java/de/westnordost/streetcomplete/screens/main/map/MapFragment.kt#L159-L160

The default values for the gesture detectors were (on my phone - not sure where/how they are set)
- shove: 43.5px
- zoom: 18.9px
- rotate: 3°

Would anyone who is interested in seeing this issue solved experiment with adjusting these thresholds so that it feels more natural and not prone to accidentally trigger the zoom gesture even though one wanted to trigger the rotate gesture (first)?

**Comment by laundmo** [comment]
Even though you keep mentioning OrganicMaps as a source for this "feature", it doesn't change that allowing a rotate to start while zooming has been default in StreetComplete for ages.

I still maintain that the default should be this, especially as its common that a combination of zooming and rotation is whats necessary to display quests hiden behind others and people are uaed to it.

I beg you, at least consider making a setting "disallow rotation while zooming" or the inverse "allow rotation while zooming".

Just consider that everyone with a opinion, even in the issue 3 years ago, has been in favour of making this a setting.

**Comment by westnordost** [comment]
Alright, I did experiment with changing the rotate threshold myself now.

---

I did not find a threshold that would result in always doing what I wanted to do (zoom or rotate). Best value with lowest error was at about 1.5°-2° for me (default is 3°). Especially a sloppy rotation or rotation with zoom needed a really low threshold to always do what I wanted - too low for the zoom-only to always trigger when I wanted it to trigger.

---

The other thing I tried out was to enable rotate while scaling but increase the threshold at which rotate gestures are detected from 3° to 5°. Didn't really work out. Accidental rotation while zooming was about as frequent as before plus there is a noticeable jump when it finally rotates the 5°+.

---

Finally, I thought I should restore the behavior like in tangram again and tried it out a bit.
And I have to admit, I was hammering that compass button to return to North-up after every zoom. Because, virtually on every zoom, the map would also rotate just a tiny bit. I think I pressed the compass button earlier all the time, too, but now being accustomed to accidental rotations on zoom virtually not happening anymore with v59, I was conscious of it.
Reverting to the old behavior turned out super annoying for me. 

---

In conclusion, I will just adjust the minimum angle for the rotate gesture to be detected to 1.5° (half that of the default) and leave the current behavior in place.

Given that the people who were most annoyed about the new default behavior are not SC users anyway and there is already a ticket in SCEE to introduce a setting, I'll close this again.

**Comment by mnalis** [comment]
For those interested in trying out; SCEE PR https://github.com/Helium314/SCEE/pull/732 now has debug `.apk` available, which allows one to choose if they want rotate disabled while zooming, and also to change a rotate threshold. 

For example; I find that _enabling_ rotation while already in zoom gesture, but _increasing rotate threshold_ hugely (e.g. to ~30 degrees), suits me the best.

---

**Update:** in the meantime, those two UI-tuning options have landed in [SCEE release 60.2](https://github.com/Helium314/SCEE/releases/tag/v60.2)

**Comment by hamishmb** [comment]
I really don't see why adding a checkbox to allow rotating when zooming was started first hasn't been done - it would surely make everyone happy.

In my case, I frequently accidentally zoom and then can't rotate the map, and I find it really annoying.

**Comment by mnalis** [comment]
> I really don't see why adding a checkbox to allow rotating when zooming was started first hasn't been done - it would surely make everyone happy.

StreetComplete is ***very*** averse to adding new preferences. Basic idea is that defaults should be good enough for (almost) everyone instead, and users should have an absolute minimum of options to tinker with. (And yes, there are reasons for that).

> In my case, I frequently accidentally zoom and then can't rotate the map, and I find it really annoying.

Have you tried [SCEE](https://wiki.openstreetmap.org/wiki/SCEE) (StreetComplete _"Expert edition"_ fork) mentioned above? Despite its name, by default it looks and functions almost exactly as StreetComplete, and it is only when you start enabling more and more of its options that it starts to noticeably divert. And it has options for tuning exactly those options which you want @hamishmb :

![Image](https://github.com/user-attachments/assets/9bcadae3-ec25-4a59-a98a-0a16f1dd4cbd)

(you can also install it in parallel to regular StreetComplete to check how it works out for you)

**Comment by hamishmb** [comment]
Fair enough and thanks for the tip, I shall give it a try.
